src/ui/ui_settings/ui_settings.js

```diff
@@ -0,0 +1,119 @@
+import { defaultsDeep } from 'lodash';
+import Bluebird from 'bluebird';
+
+import { getDefaultSettings } from './defaults';
+
+function hydrateUserSettings(user) {
+  return Object.keys(user)
+    .map(key => ({ key, userValue: user[key] }))
+    .filter(({ userValue }) => userValue !== null)
+    .reduce((acc, { key, userValue }) => ({ ...acc, [key]: { userValue } }), {});
+}
+
+function assertRequest(req) {
+  if (
+    !req ||
+    typeof req !== 'object' &&
+    typeof req.path !== 'string' &&
+    typeof req.headers !== 'object'
+  ) {
+    throw new TypeError('all uiSettings methods must be passed a hapi.Request object');
+  }
+}
+
+export class UiSettings {
+  constructor(server, status) {
+    this._server = server;
+    this._status = status;
+  }
+
+  getDefaults() {
+    return getDefaultSettings();
+  }
+
+  // returns a Promise for the value of the requested setting
+  async get(req, key) {
+    assertRequest(req);
+    return this.getAll(req)
+      .then(all => all[key]);
+  }
+
+  async getAll(req) {
+    assertRequest(req);
+    return this.getRaw(req)
+      .then(raw => Object.keys(raw)
+        .reduce((all, key) => {
+          const item = raw[key];
+          const hasUserValue = 'userValue' in item;
+          all[key] = hasUserValue ? item.userValue : item.value;
+          return all;
+        }, {})
+      );
+  }
+
+  async getRaw(req) {
+    assertRequest(req);
+    return this.getUserProvided(req)
+      .then(user => defaultsDeep(user, this.getDefaults()));
+  }
+
+  async getUserProvided(req, { ignore401Errors = false } = {}) {
+    assertRequest(req);
+    const { callWithRequest, errors } = this._server.plugins.elasticsearch.getCluster('admin');
+
+    // If the ui settings status isn't green, we shouldn't be attempting to get
+    // user settings, since we can't be sure that all the necessary conditions
+    // (e.g. elasticsearch being available) are met.
+    if (this._status.state !== 'green') {
+      return hydrateUserSettings({});
+    }
+
+    const params = this._getClientSettings();
+    const allowedErrors = [errors[404], errors[403], errors.NoConnections];
+    if (ignore401Errors) allowedErrors.push(errors[401]);
+
+    return Bluebird
+      .resolve(callWithRequest(req, 'get', params, { wrap401Errors: !ignore401Errors }))
```

Comment:
Doesn't `callWithRequest` already return a promise? Is `Bluebird.resolve` needed?